src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Lower.java

```diff
@@ -2905,6 +2908,187 @@ public void visitTypeCast(JCTypeCast tree) {
         result = tree;
     }
 
+    public void visitTypeTest(JCInstanceOf tree) {
+        if (tree.expr.type.equals(syms.objectType) && tree.pattern.type.isPrimitive()) {
+            // Object v = ...
+            // v instanceof float
+            // =>
+            // v instanceof Float
+            result = make.at(tree.pos()).TypeTest(tree.expr, make.Type(types.boxedClass(tree.pattern.type).type)).setType(syms.booleanType);
+        }
+        else if (!(tree.expr.type.isNullOrReference() && tree.pattern.type.isReference())) {
+            JCExpression exactnessCheck = null;
+
+            // translate tree.expr to resolve potential statically qualified names, etc
+            JCExpression instanceOfExpr = translate(tree.expr);
+
+            // We regard Wrapper instanceof p as unconditional if the underlying primitive of Wrapper is unconditional to p.
+            // However, we still need to emit a null check.
+            // This branch covers true unconditionality for the underlying type as well.
+            if (types.checkUnconditionallyExact(tree.expr.type, tree.pattern.type) &&
+                !(tree.expr.type.isReference() && types.isExactPrimitiveWidening(types.unboxedType(tree.expr.type), tree.pattern.type))) {
```

Comment:
Not super sure I get this line: if the expression type is unconditionally exact for the pattern type, shouldn't the test be always `true` ? What is the role of the extra guard after the `&&` ? Also, doesn't unconditionally exact implies that there is convertibility?

Reply:
Agree to all. Convertibility was covered.